tests/components/modbus/test_init.py

```diff
@@ -884,7 +884,7 @@ async def test_stop_restart(
 
     caplog.set_level(logging.INFO)
     entity_id = f"{SENSOR_DOMAIN}.{TEST_ENTITY_NAME}".replace(" ", "_")
-    assert hass.states.get(entity_id).state == STATE_UNKNOWN
+    assert hass.states.get(entity_id).state in (STATE_UNKNOWN, STATE_UNAVAILABLE)
```

Comment:
Can't we make sure that the entity has some other state that is not susceptible for the rave condition. We don't seem to need an unknown or unavailable state at this point as we're setting the state directly in the test on the next line.

Reply:
Normally it has, it's only when the test run is delayed. The test is basically not needed, but I have had situation where the value from a previous test have survived suppose it is restore_state, that kicked in.

So I can remove the assert completely if you prefer, and run the risk of what I have seen earlier, if you prefer ?

Reply:
Can we make sure that the entity has some other state?

Reply:
I would not know how, the production code run "async_add_entities(sensors)",  in case of a test, this happens as part of the setup. Then normally the test executes directly after setup, but it can happen that CI makes waits, which gives "async_update" time to run....so the only way I see if the prohibit "async_update" from running, but that means adding test code in the production code, something which I see as bad.

If I set the value in test setup, it will be overwritten if "async_update" runs so it does not help.

The only solution I see is to remove the test.

Reply:
Test removed, which should solve your comment, please approve if it is so.

Reply:
I was thinking we could make sure that `async_update` runs and sets a state that we want before we continue with the rest of the test.

Reply:
I would not know how, without changing the whole purpose of the test. The assert is gone, so I assume you have no further problem.

The assert was there for my local environment not for CI.

Reply:
It's good practice to check that the integration has created the expected state before we tell the integration to modify that state. I think we should keep this check.

Reply:
Now you confuse me,for 2 reasons: we do not have that check in other tests, and you did not like the assert with the 2 possibilities..... 

I can reinsert the test, but I cannot solve your primary problem without changing the purpose of the test, 

I have made 2 suggestions, my preference test the 2 possible values or remove the assert.

so What do you want ?

Reply:
Because that calls for a wait in the test itself, something that is not relevant for the test and thus confusing

Secondly the parametrize needs to be amended with return value and the configuration expanded accordingly....all items not relevant for the test and thus confusing.

I believe the parameters for a test shall reflect what you want to test, and the test itself shall in a minimalistic way control the test. Adding a lot of other stuff is surely possible but defies the purpose of a well defined test.

Reply:
If you add all these extras, it looks like as if there was a reason for adding it, and that you wanted to test another situation.

Reply:
It's very common in our tests to wait for tasks to finish since we use asyncio in Home Assistant. I don't see the problem with waiting. We also have test helpers to move time forward artificially.

Setting up the test appropriately to have a baseline is standard procedure. This includes making sure that an entity state has an expected value before we cause some change that should result in an entity state change.

Reply:
Seems I did not explain myself. it does wait already for the integration to be set up, I was talking about waiting for the async_update to be called, which is completely irrelevant for this test.

And do not forget all the extra pytest.mark.paramterize I need. to add, you cannot argue that it makes te test better or clearer to understand!

Seems to me, you are requesting something that really complicates this test and really for avoiding a test of 2 values instead of just one.....is that really worth or worth the discussion?

Reply:
Seems I did not explain myself. it does wait already for the integration to be set up, I was talking about waiting for the async_update to be called, which is completely irrelevant for this test.

And do not forget all the extra pytest.mark.paramterize I need. to add, you cannot argue that it makes te test better or clearer to understand!

Seems to me, you are requesting something that really complicates this test and really for avoiding a test of 2 values instead of just one.....is that really worth or worth the discussion?

Reply:
Yes. It's not good practice to directly modify the entity feature we want to test when the integration should setup and change that feature in the test. And we want to make sure that there are no tasks running that would modify the state we want to test before we cause the change (the service call in this case) that should make the integration change state.

Reply:
That is secured in our CI, but my local environment is different...that was the reason I removed the test.

The assert is not needed for our CI !

In my local environment I test a library version that allows test cases to run in parallel and production to execute several requests in parallel (now there is a lock to secure singularity). The only problem I had was exactly the test in question, hence the assert. However I have no problem living with or without that assert, I do have a problem complicating a simple test !

So the choice is still 
- assert with 2 perfectly legal values
- no assert (as in ALL the other tests)

Please choose, we cannot continue discussing this it's a waste of time.